### PR TITLE
Merkle proofs

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
Added GetProof() and VerifyProof() methods for the trie
Removed some methods from the trie interface, as they were duplicated in StorageManager.

### CODE REVIEW

File: data/interface.go [hunk 1]
```diff
@@ -151,28 +151,22 @@ type Trie interface {
 	Get(key []byte) ([]byte, error)
 	Update(key, value []byte) error
 	Delete(key []byte) error
-	Root() ([]byte, error)
+	Root() ([]byte, error) //TODO change to RootHash()
```
Comment (iulianpascalau) [inline data/interface.go:154]: Remove TODO?

File: data/trie/branchNode.go [hunk 2]
```diff
@@ -778,3 +778,18 @@ func (bn *branchNode) getAllHashes(db data.DBWriteCacher) ([][]byte, error) {
 
 	return hashes, nil
 }
+
+func (bn *branchNode) verifyProof(key []byte) (bool, []byte, []byte) {
```
Comment (iulianpascalau) [inline data/trie/branchNode.go:782]: Maybe rename this to getEncodedBytesAndHash? as this method does not verify the proof. Valid on all 3 types of nodes.

File: data/trie/patriciaMerkleTrie.go [hunk 3]
```diff
@@ -455,32 +443,6 @@ func (tr *patriciaMerkleTrie) SetNewHashes(newHashes data.ModifiedHashes) {
 	tr.newHashes = newHashes
 }
 
-// SetCheckpoint adds the current state of the trie to the snapshot database
-func (tr *patriciaMerkleTrie) SetCheckpoint(rootHash []byte) {
```
Comment (iulianpascalau) [inline data/trie/patriciaMerkleTrie.go:459]: Separation of concerns :+1:

File: data/trie/patriciaMerkleTrie.go [hunk 4]
```diff
@@ -674,7 +619,83 @@ func logMapWithTrace(message string, paramName string, hashes data.ModifiedHashe
...
+	tr.mutOperation.Lock()
+	defer tr.mutOperation.Unlock()
+
+	wantHash, err := tr.getRootHash()
+	if err != nil {
+		return false, err
+	}
+
+	key = keyBytesToHex(key)
+	for i := range proof {
```
Comment (iulianpascalau) [inline data/trie/patriciaMerkleTrie.go:669]: suggestion to change to 
```GO
for _, encodedNode := range proof {
```

**Review by iulianpascalau — COMMENTED** [review]
Good implementation! :+1: 

Suggestion to add a semi-integration test that builds up a trie with random data (couple of hundred leafs) and call a few GetProof and VerifyProof. If that test fail, maybe dump the trie contents as to better debug that edge-case.

File: data/interface.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply data/interface.go:154]: Done.

File: data/trie/branchNode.go (hunk 2 above)
Comment (BeniaminDrasovean) [reply data/trie/branchNode.go:782]: Renamed to getNextHashAndKey.

File: data/trie/patriciaMerkleTrie.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply data/trie/patriciaMerkleTrie.go:669]: Done.

File: data/trie/patriciaMerkleTrie_test.go [hunk 5]
```diff
@@ -780,6 +781,49 @@ func TestPatriciaMerkleTrie_VerifyProofFromDifferentTrieShouldNotWork(t *testing
...
+
+	tr := emptyTrie()
+	hsh := keccak.Keccak{}
+
+	nrLeaves := 10000
+	values := make([][]byte, nrLeaves)
+	numRuns := 5000
+
+	for i := 0; i < nrLeaves; i++ {
+		randNum := rand.Intn(100000000)
```
Comment (iulianpascalau) [inline data/trie/patriciaMerkleTrie_test.go:795]: Can remain as it is. I would have done it like this:
```GO
buff := make([]byte, 32)
_, _ = rand.Read(buff) 
//now buff will have 32 random bytes, rand should by crypto/rand
```

**Review by iulianpascalau — APPROVED** [review]

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: data/trie/leafNode.go [hunk 6]
```diff
@@ -374,3 +374,19 @@ func (ln *leafNode) getAllHashes(_ data.DBWriteCacher) ([][]byte, error) {
 
 	return [][]byte{ln.hash}, nil
 }
+
+func (ln *leafNode) getNextHashAndKey(key []byte) (bool, []byte, []byte) {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/trie/leafNode.go:378]: Added.

File: data/trie/extensionNode.go [hunk 7]
```diff
@@ -631,3 +631,18 @@ func (en *extensionNode) getAllHashes(db data.DBWriteCacher) ([][]byte, error) {
 
 	return hashes, nil
 }
+
+func (en *extensionNode) getNextHashAndKey(key []byte) (bool, []byte, []byte) {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/trie/extensionNode.go:635]: Added.

File: data/trie/branchNode.go [hunk 8]
```diff
@@ -778,3 +778,18 @@ func (bn *branchNode) getAllHashes(db data.DBWriteCacher) ([][]byte, error) {
 
 	return hashes, nil
 }
+
+func (bn *branchNode) getNextHashAndKey(key []byte) (bool, []byte, []byte) {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply data/trie/branchNode.go:782]: Added.

**Review by iulianpascalau — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

**Review by mateigraura — APPROVED** [review]
system test passed
